File: medium/2906_construct_product_matrix/solution.py
from typing import List


class Solution:
    def constructProductMatrix(self, grid: List[List[int]]) -> List[List[int]]:
        rows = len(grid)
        cols = len(grid[0])

        product_grid = []
        for _ in range(rows):
            product_grid.append([1] * cols)

        for r1 in range(rows):
            for c1 in range(cols):
                for r2 in range(rows):
                    for c2 in range(cols):
                        product_grid[r1][c1] *= grid[r2][c2]

                product_grid[r1][c1] = (
                    product_grid[r1][c1] // grid[r1][c1]
                ) % 12345

        return product_grid

File: medium/2906_construct_product_matrix/solution.cpp
#include <vector>

class Solution {
public:
  std::vector<std::vector<int>>
  constructProductMatrix(std::vector<std::vector<int>> &grid) {
    const int rows = grid.size();
    const int cols = grid[0].size();

    std::vector<std::vector<int>> productGrid(rows, std::vector<int>(cols, 1));

    int row, col;

    long prefixProduct = 1;
    for (row = 0; row < rows; ++row) {
      for (col = 0; col < cols; ++col) {
        productGrid[row][col] = prefixProduct;
        prefixProduct = (prefixProduct * grid[row][col]) % 12345;
      }
    }

    long suffixProduct = 1;
    for (row = rows - 1; row > -1; --row) {
      for (col = cols - 1; col > -1; --col) {
        productGrid[row][col] = (productGrid[row][col] * suffixProduct) % 12345;
        suffixProduct = (suffixProduct * grid[row][col]) % 12345;
      }
    }

    return productGrid;
  }
};
